ambari-server/src/main/java/org/apache/ambari/server/topology/addservice/StackAdvisorAdapter.java

```diff
@@ -113,47 +115,128 @@ AddServiceInfo recommendLayout(AddServiceInfo info) {
-      keepNewServicesOnly(recommendedLayout, info.newServices());
+      // Validate layout
+      Map<String, Set<String>> recommendedComponentHosts = getComponentHostMap(recommendedLayout);
+      StackAdvisorRequest validationRequest = request.builder()
+        .forHostsGroupBindings(response.getRecommendations().getBlueprintClusterBinding().getHostgroupHostMap())
+        .withComponentHostsMap(recommendedComponentHosts)
+        .forHostComponents(getHostComponentMap(recommendedComponentHosts)).build();
+      validate(validationRequest);
 
-      return info.withNewServices(recommendedLayout);
+      Map<String,Map<String,Set<String>>> newServiceRecommendations = keepNewServicesOnly(recommendedLayout, info.newServices());
+      LayoutRecommendationInfo recommendationInfo = new LayoutRecommendationInfo(
+        response.getRecommendations().getBlueprintClusterBinding().getHostgroupHostMap(),
+        recommendedLayout);
+      return info.withLayoutRecommendation(newServiceRecommendations, recommendationInfo);
     }
     catch (AmbariException|StackAdvisorException ex) {
       throw new IllegalArgumentException("Layout recommendation failed.", ex);
     }
   }
 
-  Set<ValidationResponse.ValidationItem> validateRecommendedLayout(StackId stackId,
-                                                              Map<String,Map<String,Set<String>>> recommendedLayout,
-                                                              Map<String, Set<String>> recommendedHostgroups) throws StackAdvisorException {
-    Map<String, Set<String>> componentsToHosts = getComponentHostMap(recommendedLayout);
-    Map<String, Set<String>> hostsToComponents = getHostComponentMap(componentsToHosts);
-    List<String> hosts = ImmutableList.copyOf(hostsToComponents.keySet());
-
-    StackAdvisorRequest request = StackAdvisorRequest.StackAdvisorRequestBuilder
-      .forStack(stackId)
-      .ofType(StackAdvisorRequest.StackAdvisorRequestType.HOST_GROUPS)
-      .forHosts(hosts)
-      .forServices(recommendedLayout.keySet())
-      .forHostComponents(hostsToComponents)
-      .forHostsGroupBindings(recommendedHostgroups)
-      .withComponentHostsMap(componentsToHosts)
-      .withGPLLicenseAccepted(serverConfig.getGplLicenseAccepted())
-      .build();
-    ValidationResponse response = stackAdvisorHelper.validate(request);
-
-    return response.getItems();
+  AddServiceInfo recommendConfigurations(AddServiceInfo info) {
+    Configuration config = info.getConfig();
+
+    if (info.getRequest().getRecommendationStrategy().shouldUseStackAdvisor()) {
+      // Reuse information from layout recommendation.
+      // Layout recommendation is currently mandatory. When it will be optional, this will have to be rewritten to
+      // compute this information if missing
+      LayoutRecommendationInfo layoutInfo =
+        info.getRecommendationInfo().orElseThrow(() -> new IllegalStateException("Config recommendation must happen after layout recommendation"));
```

Comment:
Layout recommendation is no longer mandatory, it is skipped with `"components"`-only input (#2714).